### PR TITLE
add query/node/time metrics to DirectDruidClient

### PR DESCRIPTION
**Pull request by xvrl**
This is meant to replace what was removed in #1467 assuming performance is not affected

### CODE REVIEW

**Comment by xvrl** [comment]
if this looks good, I'll add docs

File: server/src/main/java/io/druid/client/DirectDruidClient.java [hunk 1]
```diff
@@ -152,9 +159,16 @@ public int getNumOpenConnections()
     try {
       log.debug("Querying url[%s]", url);
 
+      final long requestStartTime = System.currentTimeMillis();
+
+      final ServiceMetricEvent.Builder builder = toolChest.makeMetricBuilder(query);
+      builder.setDimension("server", host);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: is "server" not canonized in a constant somewhere?

Comment (drcrallen) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: Oh nvmnd, that's different server.

File: server/src/main/java/io/druid/client/DirectDruidClient.java [hunk 2]
```diff
@@ -163,7 +177,9 @@ public int getNumOpenConnections()
         public ClientResponse<InputStream> handleResponse(HttpResponse response)
         {
           log.debug("Initial response from url[%s]", url);
-          startTime = System.currentTimeMillis();
+          responseStartTime = System.currentTimeMillis();
+          emitter.emit(builder.build("query/node/ttfb", responseStartTime - requestStartTime));
```
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: what does ttfb stand for? is there a reason to not just spell it out?

Comment (xvrl) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: `ttfb` stands for "time to first byte". I could spell it out, or does someone has a better suggestion?

**Comment by cheddar** [comment]
:+1:

File: server/src/main/java/io/druid/client/DirectDruidClient.java (hunk 2 above)
Comment (cheddar) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: `ttfb` is pretty standard amongst internet performance measurements.  I vote to leave it.

**Comment by fjy** [comment]
:+1:

File: server/src/main/java/io/druid/client/DirectDruidClient.java (hunk 2 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/client/DirectDruidClient.java]: Allrighty works for me

**Comment by xvrl** [comment]
ok, let me add docs then

**Comment by drcrallen** [comment]
It would be nice, however, if this is not merged until a perf test is run on it.

**Comment by xvrl** [comment]
Yes, we should definitely do some performance testing before merging.

**Comment by himanshug** [comment]
+1

**Comment by himanshug** [comment]
@drcrallen did you identify the slowness cause for #1467

also looking at HttpPostEmitter.emit(..) which is potentially the reason for slowness. I think first synchronized at line-142 can be removed without any impact on the behavior.
And, second synchronized at line-166 can be removed with the understanding that in the worst case of concurrent calls we might end up buffering more events than configured. so config.getMaxBufferSize() becomes more of a hint than a strict limit. In this case, that looks ok to me than compromising the performance of druid in potentially many places where metrics are being emitted.

**Comment by nishantmonu51** [comment]
+1

**Comment by drcrallen** [comment]
@himanshug I did not identify the exact bottleneck other than "somewhere in the http emitter"

**Comment by fjy** [comment]
@xvrl @drcrallen are we able to merge this?

**Comment by drcrallen** [comment]
@fjy Gathering data right now. Should know today sometime.

**Comment by drcrallen** [comment]
Ok, performance tests are coming back fine.

**Pull request merged**